File: 946.cpp
#include <stack>

struct Solution {
    bool validateStackSequences(const vector<int>& pushed, const vector<int>& popped) {
        stack<int> st;
        size_t pushPtr = 0, popPtr = 0, pushSize = pushed.size(), popSize = popped.size();
        while (true){
            while (popPtr < popSize && !st.empty() && popped[popPtr] == st.top()){
                st.pop();
                ++popPtr;
            }
            if (popPtr == popSize){
                return true;
            }
            if (pushPtr < pushSize){
                st.push(pushed[pushPtr++]);
            } else {
                return false;
            }
        }
    }
};

File: sol/946/946.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = list()
        pushPtr = 0
        popPtr = 0
        while True:
            while popPtr < len(popped) and len(stack) != 0 and popped[popPtr] == stack[-1]:
                stack.pop()
                popPtr += 1
            if popPtr == len(popped):
                return True
            if pushPtr < len(pushed):
                stack.append(pushed[pushPtr])
                pushPtr += 1
            else:
                return False

